### ISSUE TITLE
Sequence-to-sequence LSTM encoder-decoder with MaskZero

### ISSUE THREAD

**Issue by grachev** [Element-Research/rnn#199]
Hi, I have a question about using variable length input and output sequences in minibatch mode. I am observing that if I increase the length of the padding, the training does not only become slower, but it also does not converge as well, which makes me suspect that I am doing something wrong. Here is my code for model and criterion. I would greatly appreciate answers to the following questions:

1. Is this the correct way to implement such model?
2. I am padding all sequences (encoder input, encoder output and decoder output) with zeros from the left. For encoder output I add GO token after padding (and before real sequence), for decoder output I add EOS token after real sequence end. Am I doing this correctly?

```lua
-- Encoder
local enc = nn.Sequential()
enc:add(nn.LookupTableMaskZero(opt.vocabSize, opt.hiddenSize))
enc:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
local encLSTM = nn.LSTM(opt.hiddenSize, opt.hiddenSize):maskZero(1)
local encLSTM2 = nn.LSTM(opt.hiddenSize, opt.hiddenSize):maskZero(1)
enc:add(nn.Sequencer(encLSTM))
enc:add(nn.Sequencer(encLSTM2))
enc:add(nn.SelectTable(-1))

-- Decoder
local dec = nn.Sequential()
dec:add(nn.LookupTableMaskZero(opt.vocabSize, opt.hiddenSize))
dec:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
local decLSTM = nn.LSTM(opt.hiddenSize, opt.hiddenSize):maskZero(1)
local decLSTM2 = nn.LSTM(opt.hiddenSize, opt.hiddenSize):maskZero(1)
dec:add(nn.Sequencer(decLSTM))
dec:add(nn.Sequencer(decLSTM2))
dec:add(nn.Sequencer(nn.MaskZero(nn.Linear(opt.hiddenSize, opt.vocabSize), 1)))
dec:add(nn.Sequencer(nn.MaskZero(nn.LogSoftMax(), 1)))

local criterion = nn.SequencerCriterion(nn.MaskZeroCriterion(nn.ClassNLLCriterion(),1))
```

**Comment by nicholas-leonard** [comment]
Hi @grachev 

1. Yes. Your model seems correct.

2. I think the padding should go add the end of the decoder. Otherwise when you set prevOutput and prevCell it basically gets ignored. So encoder would be : `{0,0,0,GO,1,2,3,4}` and decoder would be `{1,2,3,4,EOS,0,0,0}`.

Let me know if that works better.

**Comment by grachev** [comment]
Yes, that was exactly it thanks! Maybe it is worth adding this to the example, since a lot of people will be using this in batch mode with variable-length sequences.

**Comment by nicholas-leonard** [comment]
@grachev Could you send a Pull Request with these changes added to the example?
